controllers/drain_controller.go

```diff
@@ -0,0 +1,461 @@
+		err = dr.Get(ctx, client.ObjectKey{Name: nodeObj.GetName(), Namespace: vars.Namespace}, snns)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				reqLogger.V(2).Info("node doesn't have a sriovNetworkNodePolicy")
+				continue
+			}
+			return nil, err
+		}
+
+		if snns.GetName() == node.GetName() {
+			currentSnns = snns.DeepCopy()
+		}
+
+		if utils.ObjectHasAnnotation(snns, constants.NodeStateDrainAnnotationCurrent, constants.Draining) ||
+			utils.ObjectHasAnnotation(snns, constants.NodeStateDrainAnnotationCurrent, constants.DrainComplete) {
+			current++
+		}
+	}
+	reqLogger.Info("Max node allowed to be draining at the same time", "MaxParallelNodeConfiguration", maxUnv)
+	reqLogger.Info("Count of draining", "drainingNodes", current)
+
+	// if maxUnv is zero this means we drain all the nodes in parallel without a limit
+	if maxUnv == -1 {
+		reqLogger.Info("draining all the nodes in parallel")
+	} else if current >= maxUnv {
+		// the node requested to be drained, but we are at the limit so we re-enqueue the request
+		reqLogger.Info("MaxParallelNodeConfiguration limit reached for draining nodes re-enqueue the request")
+		// TODO: make this time configurable
+		return &reconcile.Result{RequeueAfter: 5 * time.Second}, nil
+	}
+
+	if currentSnns == nil {
+		return nil, fmt.Errorf("failed to find sriov network node state for requested node")
+	}
+
+	err = utils.AnnotateObject(currentSnns, constants.NodeStateDrainAnnotationCurrent, constants.Draining, dr.Client)
+	if err != nil {
+		reqLogger.Error(err, "failed to annotate node with annotation", "annotation", constants.Draining)
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (dr *DrainReconcile) findNodePoolConfig(ctx context.Context, node *corev1.Node) (*sriovnetworkv1.SriovNetworkPoolConfig, []corev1.Node, error) {
+	logger := log.FromContext(ctx)
+	logger.Info("findNodePoolConfig():")
+	// get all the sriov network pool configs
+	npcl := &sriovnetworkv1.SriovNetworkPoolConfigList{}
+	err := dr.List(ctx, npcl)
+	if err != nil {
+		logger.Error(err, "failed to list sriovNetworkPoolConfig")
+		return nil, nil, err
+	}
+
+	selectedNpcl := []*sriovnetworkv1.SriovNetworkPoolConfig{}
+	var defaultNpcl *sriovnetworkv1.SriovNetworkPoolConfig
+
+	for _, npc := range npcl.Items {
+		// we skip hw offload objects
+		if npc.Spec.OvsHardwareOffloadConfig.Name != "" {
+			continue
+		}
+
+		// we save the default to use it if we don't find any other one
+		if npc.GetName() == "default" {
+			defaultNpcl = npc.DeepCopy()
+			continue
+		}
+
+		// if the node selector is empty we skip
+		if npc.Spec.NodeSelector == nil {
+			continue
+		}
+
+		selector, err := metav1.LabelSelectorAsSelector(npc.Spec.NodeSelector)
+		if err != nil {
+			logger.Error(err, "failed to create label selector from nodeSelector", "nodeSelector", npc.Spec.NodeSelector)
+			return nil, nil, err
+		}
+
+		// If a pool with a nil or empty selector creeps in, it should match nothing, not everything.
```

Comment:
empty selector selects everything usually.

moreover for the default object it technically select everything.

what if i want to override the default obj ?

Reply:
I removed the default object as requested :)

Reply:
You removed creation of the default Pool from the main.go, but you still have logic to recreated default pool in the https://github.com/SchSeba/sriov-network-operator-1/blob/8d80f3ea5db2536bb7bebf3773a50181279b8dc7/controllers/sriovnetworkpoolconfig_controller.go#L69. Do we want to remove this as well?

Reply:
@SchSeba please address this comment

Reply:
you are right thanks!

Reply:
Thanks @SchSeba 

regarding

> empty selector selects everything usually.

can we make this behaviour aligned with how kubernetes handles empty selectors ? (match all)
currently the opposite happens.

Reply:
+1 to change the behavior to "match all" for an empty selector unless we have a good reason not to do so

Reply:
done :)